run.c

```diff
@@ -570,7 +577,7 @@ int main(int argc, char *argv[]) {
         // following BOS token (1), sentencepiece decoder strips any leading whitespace (see PR #89)
         char *token_str = (token == 1 && vocab[next][0] == ' ') ? vocab[next]+1 : vocab[next];
         printf("%s", token_str);
-        fflush(stdout);
+        if (bufferflush==pos && strlen(outbuff)<=4096) { fflush(stdout); bufferflush+=buffertokens; } // flush after every n tokens
```

Comment:
1. outbuf is not necessarily zero terminated
2. flush at the end 

```suggestion
        if (bufferflush==pos || pos == steps) { fflush(stdout); bufferflush+=buffertokens; } // flush after every n tokens
```

Reply:
Logic:

1. On certain systems, writing to console / screen is **relatively and variably** slow compared to the computation.
2. Currently each token is written to screen and flushed so **inherits 1**
3. We **speed up by batching** the number of tokens to be flushed to screen.
4. We enable the **number of tokens** to be batched as a **argument** for **flexibility** 
5. We need to **size a appropriate buffer**, so that we do not encounter a buffer overflow
6. **We do not know beforehand** the batch size, the size of each token or the context size 
7. We estimate a **safe buffer size** like this: **context size** * (**average token length** + **safe margin**)
8. Context size is determined from the model **config.seq_len** parameter.  
9. BPE subword units usually are **2-6 chars**. Better to err, I used **6 as avg token length** and **2 for a safe margin.**
10. That **explicitly sized buffer** and **its calculated size** is passed as parameters to setvbuf to buffer the output. setvbuf will round it up to the next power of 2.
11. We are **not** requesting setvbuf to automatically create a buffer for us as with the NULL parameter **nor** are we passing BUFSIZ as our **purpose** is a buffer that can hold **context size** * (**average token length** + **safe margin**) for the maximum case such as **batch size = context size**.
12. Once that is set, **each time the buffer is filled with a batch** of the required number of tokens, **it is flushed**. 

Your suggestion were helpful for latest commit. These are my findings:

``if (bufferflush==pos && strlen(outbuff)<=4096)`` , here the ``&& strlen(outbuff)<=4096`` check is **not required and wrong**. Also reading a buffer managed by setvbuf is **undefined behaviour**. 

``if (bufferflush==pos || pos == steps)``, here ``|| pos == steps`` would not be needed as the **buffer can already hold max steps** (config.seq_len) now.

**You were right about a bug.**

**The bug was a hard coded buffer size** that did not take into account **context size** * (**average token length** + **safe margin**). Even the increase to 4096 would have caused the overflow on Meta Llama2 7b.

**Thank you.**

Reply:
Okay this doesn't build with MSVC. Need to figure an alternative.

Reply:
"The default buffer size BUFSIZ is expected to be the most efficient buffer size for file I/O on the implementation", you need at least some benchmarks to second-guess it. For purity, safe buffer size would be context size * (**max** token length). There is no guarantee that it's smaller than 4096, which is an unnamed constant you use multiple times. But it's not clear that such purity is needed. In fact, I am not sure even sure that flushing after max tokens vs max bytes (which would have a much simpler implementation) is a right approach.

My proposed `|| pos == steps` is there to flush at the end of the loop, although I just realized it's not needed, sorry about that, because of:
https://github.com/karpathy/llama2.c/blob/af8708d87bcda7fda97b93f6c135dd43ea78106c/run.c#L584

Reply:
If you have benchmarks on BUFSIZ please share it here. It would be of great help! Also kindly test with BUFSIZE and steps 0 and any buffer_tokens parameters such as 4, 8, 16, 32, 64 ... .

Usage:

run <checkpoint_file> [temperature] [steps] [prompt] [buffer_tokens]

Where buffer_tokens is the number of tokens to be buffered.